CA-80 Add tests for /faq endpoint

edvinbasil opened the pull request.

edvinbasil committed:
Add tests for GET /faq endpoint

- Prepopulating test db to check /faq/:id
- closing db connection after each tests to prevent opn handles.

edvinbasil committed:
Add tests for POST /faq

edvinbasil committed:
Add tests for DELETE /faq

spiray commented on src/api/faq/faq.test.js:
```diff
@@ -1,11 +1,112 @@
+
+const ids = {
+  invalid: 'invalid_id',
+  nonexistant: 'aaaaaaaaaaaaaaaaaaaaaaaa',
```
nonexistant -> nonexistent

spiray commented on src/api/faq/faq.test.js:
```diff
@@ -1,11 +1,112 @@
+}
+
+beforeAll(prePopulate);
+afterAll(() => FAQ.deleteMany({}).then(() => mongoose.connection.close()));
```
I would use this structure to keep everything sync
```js
afterAll(async () => {
  await FAQ.deleteMany({});
  connection.close()
});
```

spiray commented on src/api/faq/faq.test.js:
```diff
@@ -1,11 +1,112 @@
 const request = require('supertest');
+const mongoose = require('mongoose');
```
```js
const { connection } = require('mongoose');
```

spiray commented on src/api/faq/faq.test.js:
```diff
@@ -1,11 +1,112 @@
+      .expect(200)
+      .then((res) => {
+        expect(res.body)
+          .toEqual(expect.any(Array));
```
I would also add a check that the array length is greater than 0. This can be in this test or a separate test.

spiray commented on src/api/faq/faq.test.js:
```diff
@@ -1,11 +1,112 @@
+      .post('/faq')
+      .send({ question: 'Test Question', answer: 'Test answer' })
+      .expect(200)
+      .then((res) => {
```
I would make this sync with async/await as well.

spiray commented:
Great work @edvinbasil! Really nicely written tests!

spiray commented:
![](https://github.trello.services/images/mini-trello-icon.png) [As a Developer, I would like to have the FAQ endpoints tested](https://trello.com/c/JouERUOX/80-as-a-developer-i-would-like-to-have-the-faq-endpoints-tested)

edvinbasil committed:
Make changes from pull request comments

- Fixed typo
- Changed from using .then to async/await
- Fixed mongoose connection import
- Add test to check array length > 0 for GET /faq

edvinbasil commented:
Made changes to resolve the issues that were brought up. @spiray

spiray commented:
Looks good! The word existent appears a few more times with the wrong spelling. The last thing I am thinking of is moving `prePopulate` to the helpers file. @ewenjo What do you think?

edvinbasil committed:
Fix typo

edvinbasil commented:
:sweat_smile: I changed the references of the word through my editor. Guess it didnt pickup the strings.
Its fixed now (double checked this time) :)

ewenjo commented:
I agree with @spiray that we should move the `prePopulate` to the helpers file. This could be used with other tests. There are a still few minor spelling mistakes. Other than that, it looks good to me.

edvinbasil commented:
Should I modify prepopulate to accept a model and and object to insert into the database?
so something like
```js
async function prePopulate(Model, document) {
  const { _id } = await new Model(document).save();
  return _id;
}
```
and then call in in beforeAll like:
```js
const question = 'Populate Question';
const answer = 'Populate answer';
const doc = { question, answer };

beforeAll(async () => {
  ids.valid = await prePopulate(FAQ, doc);
});
```

ewenjo commented:
Yeah, that looks good.

edvinbasil committed:
Seperate prePopulate to a helper file

- Change faq test file to use this helper file

edvinbasil commented:
:+1: Made the required changes

Commit:
Minor spelling fixes.

ewenjo reviewed (approved):
Lgtm, good work!
